src/wallet/silentpayments.cpp

```diff
@@ -0,0 +1,179 @@
+#include <wallet/silentpayments.h>
+#include <arith_uint256.h>
+#include <coins.h>
+#include <crypto/common.h>
+#include <crypto/hmac_sha512.h>
+#include <key_io.h>
+#include <undo.h>
+#include <logging.h>
+
+namespace wallet {
+
+Recipient::Recipient(const CKey& scan_seckey, const CKey& spend_seckey)
+{
+    m_scan_seckey = scan_seckey;
+    m_scan_pubkey = CPubKey{m_scan_seckey.GetPubKey()};
+    m_spend_seckey = spend_seckey;
+    m_spend_pubkey = CPubKey{m_spend_seckey.GetPubKey()};
+}
+
+void Recipient::ComputeECDHSharedSecret(const CPubKey& sender_public_key, const std::vector<COutPoint>& tx_outpoints)
+{
+    const auto& outpoint_hash = HashOutpoints(tx_outpoints);
+    auto tweaked_scan_seckey = m_scan_seckey.MultiplyTweak(outpoint_hash.begin());
+    CPubKey result = tweaked_scan_seckey.SilentPaymentECDH(sender_public_key);
+
+    m_ecdh_pubkey = result;
+    assert(m_ecdh_pubkey.IsValid());
+}
+
+std::pair<CKey,CPubKey> Recipient::CreateOutput(const uint32_t output_index) const
+{
+    HashWriter h;
+    h.write(Span{m_ecdh_pubkey});
+    unsigned char num[4];
+    WriteBE32(num, output_index);
+    h << num;
+    uint256 shared_secret = h.GetSHA256();
+
+    const auto& result_pubkey{m_spend_pubkey.AddTweak(shared_secret.begin())};
+    const auto& result_seckey{m_spend_seckey.AddTweak(shared_secret.begin())};
+
+    return {result_seckey, result_pubkey};
+}
+
+std::pair<CPubKey,CPubKey> Recipient::GetAddress() const
+{
+    return {m_scan_pubkey, m_spend_pubkey};
+}
+
+std::vector<CKey> Recipient::ScanTxOutputs(std::vector<XOnlyPubKey> output_pub_keys)
```

Comment:
this function should be "label aware," meaning if the wallet implements labels later on, this function does not need to change.

Reply:
Leaving this alone for now. This can be easily updated to add label support when it makes sense.